### PR TITLE
Periodic order events table clean up

### PR DESCRIPTION
**Pull request by squadgazzz**
# Description
Add a background task that periodically drops order events from DB that are very old (configurable).

# Changes

- [ ] Configurable interval and threshold of the cleaning task
- [ ] Adds a new metric in order to create a separate alerting when the clean-up wasn't successful for a long period of time

## How to test
Added integro-test

## Related Issues
Implements the first task of the #1977

### CODE REVIEW

File: crates/autopilot/src/arguments.rs [hunk 1]
```diff
@@ -203,6 +207,16 @@ pub struct Arguments {
...
     pub solve_deadline: Duration,
+
+    /// Time interval in days between each cleanup operation of the
+    /// `order_events` database table.
+    #[clap(long, env, default_value = "1", value_parser = duration_from_days)]
+    pub order_events_cleanup_interval: Duration,
+
+    /// Age threshold in days for order events to be eligible for cleanup in the
+    /// `order_events` database table.
+    #[clap(long, env, default_value = "30", value_parser = duration_from_days)]
```
Comment (squadgazzz) [inline crates/autopilot/src/arguments.rs:218]: TBD: default value

File: crates/autopilot/src/arguments.rs [hunk 2]
```diff
@@ -203,6 +207,16 @@ pub struct Arguments {
         value_parser = shared::arguments::duration_from_seconds,
     )]
     pub solve_deadline: Duration,
+
+    /// Time interval in days between each cleanup operation of the
+    /// `order_events` database table.
+    #[clap(long, env, default_value = "1", value_parser = duration_from_days)]
```
Comment (squadgazzz) [inline crates/autopilot/src/arguments.rs:213]: TBD: default value

Comment (fleupold) [reply crates/autopilot/src/arguments.rs:213]: I'm even more positive on your suggested `humanreadtime` change, as I find it dangerous to configure different time parameters in different units (given the name doesn't indicate in which unit this value is expected to be taken)

File: crates/autopilot/src/periodic_db_cleanup.rs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+                    );
+                    Metrics::get()
+                        .last_order_events_cleanup_run
+                        .set(Utc::now().timestamp())
+                }
+                Err(err) => {
+                    tracing::warn!(?err, "failed to delete order events before {}", timestamp)
+                }
+            }
+            interval.tick().await;
```
Comment (fleupold) [inline crates/autopilot/src/periodic_db_cleanup.rs:51]: til about `tokio::time::interval`. Are you sure though that the first tick doesn't return immediately (at least that's what I read from their [example](https://docs.rs/tokio/latest/tokio/time/struct.Interval.html#examples). Maybe a simple sleep would also work?

File: crates/autopilot/src/periodic_db_cleanup.rs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+            byte_array::ByteArray,
+            order_events::{OrderEvent, OrderEventLabel},
+        },
+        sqlx::PgConnection,
+    };
+
+    #[tokio::test]
+    #[ignore]
+    async fn postgres_count_rows_in_table_() {
+        let db = Postgres::new("postgresql://").await.unwrap();
```
Comment (fleupold) [inline crates/autopilot/src/periodic_db_cleanup.rs:83]: @MartinquaXD was experimenting with spawning up new databases for tests using docker. Maybe that's safer than nuking our existing local database?

File: crates/autopilot/src/periodic_db_cleanup.rs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+        let count = order_events_before_timestamp_count(
+            &mut ex,
+            latest_timestamp + chrono::Duration::days(1),
+        )
+        .await;
+        assert_eq!(count, 2u64);
+
+        let count = order_events_before_timestamp_count(&mut ex, latest_timestamp).await;
+        assert_eq!(count, 1u64);
```
Comment (fleupold) [inline crates/autopilot/src/periodic_db_cleanup.rs:114]: What does this actually test? Is it just making sure that the SQL query is correct? I'd expect us to instantiate a new instance of `OrderEventsCleaner` and run it (potentially also warping time to see it run again after the wait time)

File: crates/autopilot/src/periodic_db_cleanup.rs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+            interval.tick().await;
+        }
+    }
+}
+
+#[derive(prometheus_metric_storage::MetricStorage)]
+struct Metrics {
+    /// Timestamp of the last successful `order_events` table cleanup.
+    #[metric(name = "periodic_db_cleanup")]
+    last_order_events_cleanup_run: prometheus::IntGauge,
```
Comment (fleupold) [inline crates/autopilot/src/periodic_db_cleanup.rs:60]: One issue with using a Gauge here is that if the first run after restart fails, the Gauge will be not initialized which may cause an alert to fire (even if it ran successfully the last day and will run successfully again tomorrow). Also I'm not sure if it's easy in Grafana to compare a gauge value against now.

Maybe using a counter is enough (for this one we can alert if the rate e.g. per week is less than a threshold)?

**Review by fleupold — COMMENTED** [review]
Looks good. I'm not super happy with the test (I feel like it should create the object under test), otherwise only small comments.

File: crates/autopilot/src/periodic_db_cleanup.rs [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        let mut interval = time::interval(self.config.cleanup_interval);
+        loop {
+            let timestamp: DateTime<Utc> = Utc::now() - self.config.event_age_threshold;
+            match self.db.delete_events_before(timestamp).await {
+                Ok(affected_rows_count) => {
+                    tracing::debug!(
+                        "deleted {:?} order events before {}",
+                        affected_rows_count,
+                        timestamp
+                    );
```
Comment (MartinquaXD) [inline crates/autopilot/src/periodic_db_cleanup.rs:42]: nit: We try to have logs which have a static string and log any additional data as named variables like this:
```rust
tracing::debug!(affected_rows, timestamp, "deleted old order events");
```
This generally makes it nicer to find these logs in kibana because their string is always constant while still being able to circle in on logs with specific values when that's needed.

File: crates/autopilot/src/periodic_db_cleanup.rs (hunk 4 above)
Comment (MartinquaXD) [reply crates/autopilot/src/periodic_db_cleanup.rs:83]: The dockerized tests PR is still WIP. And in fact it looks like even in the dockerized test setup the strategy of having a single DB that gets reset after every test might be needed to not slow down our tests unnecessarily by doing DB migrations over and over again (there should be a way to the migration once and reuse the same container but is currently not possible due to a bug in the dependency; long story).

Long story short this test is indeed how we currently do DB tests and BTW we don't actually nuke the DB (although it wouldn't matter) because we are setting up a DB transaction that we don't commit.

File: database/sql/V057__add_timestamp_index_to_order_events.sql [hunk 8]
```diff
@@ -0,0 +1,2 @@
+-- Adds an index on the 'timestamp' column of 'order_events' table to facilitate efficient periodic cleanups.
+CREATE INDEX order_events_by_timestamp ON order_events (timestamp);
```
Comment (MartinquaXD) [inline database/sql/V057__add_timestamp_index_to_order_events.sql:2]: Not sure we should have an index to speed up an operation we do at most once per day (current granularity of the interval setting).

File: crates/autopilot/src/periodic_db_cleanup.rs (hunk 5 above)
Comment (MartinquaXD) [reply crates/autopilot/src/periodic_db_cleanup.rs:114]: Good point. tokio has some [helper functions](https://docs.rs/tokio/latest/tokio/time/index.html#functions) (functions of `test-util` feature) for that.

File: crates/autopilot/src/arguments.rs (hunk 2 above)
Comment (squadgazzz) [reply crates/autopilot/src/arguments.rs:213]: Will be done in a separate PR https://github.com/cowprotocol/services/issues/2085

**Review by MartinquaXD — COMMENTED** [review]
Looks good over all. Just a few nits 👍

File: crates/autopilot/src/periodic_db_cleanup.rs (hunk 3 above)
Comment (squadgazzz) [reply crates/autopilot/src/periodic_db_cleanup.rs:47]: `sleep` leads to time drifts, which is not expected behavior for the current logic. Added initial `tick` call instead.

File: crates/autopilot/src/periodic_db_cleanup.rs (hunk 5 above)
Comment (squadgazzz) [reply crates/autopilot/src/periodic_db_cleanup.rs]: Updated the test. Unfortunately couldn't find a way to use `tokio::time::advance` function. Onse `tokio::time::pause()` is called, `PgPool` fails to acquire a connection leading to timeout panic. Current approach uses much more time but doesn't have an issue with the db connection allocation.

File: crates/autopilot/src/periodic_db_cleanup.rs (hunk 6 above)
Comment (squadgazzz) [reply crates/autopilot/src/periodic_db_cleanup.rs]: Replaced with counter

File: crates/autopilot/src/periodic_db_cleanup.rs (hunk 7 above)
Comment (squadgazzz) [reply crates/autopilot/src/periodic_db_cleanup.rs]: Updated the log

File: database/sql/V057__add_timestamp_index_to_order_events.sql (hunk 8 above)
Comment (squadgazzz) [reply database/sql/V057__add_timestamp_index_to_order_events.sql]: Makes sense, removed for now

File: crates/autopilot/src/periodic_db_cleanup.rs [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+        assert_eq!(ids.len(), 3);
+        assert!(ids.contains(&event_a.order_uid));
+        assert!(ids.contains(&event_b.order_uid));
+        assert!(ids.contains(&event_c.order_uid));
+
+        let config =
+            OrderEventsCleanerConfig::new(Duration::from_millis(500), Duration::from_millis(2000));
+        let cleaner = OrderEventsCleaner::new(config, db.clone());
+
+        tokio::task::spawn(cleaner.run_forever());
```
Comment (squadgazzz) [inline crates/autopilot/src/periodic_db_cleanup.rs:123]: I am personally against the `fire and forget` pattern. I believe the thread will continue executions in the background during the execution of other tests, which could lead to unexpected results. Currently, it works because `order_events` table is not queried in any other test. One more reason for the https://github.com/cowprotocol/services/pull/2023

File: crates/autopilot/src/periodic_db_cleanup.rs (hunk 3 above)
Comment (fleupold) [reply crates/autopilot/src/periodic_db_cleanup.rs:47]: nit: why don't we tick().await inside the loop in first place instead of calling it twice?

File: crates/autopilot/src/periodic_db_cleanup.rs [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+        assert!(ids.contains(&event_c.order_uid));
+
+        let config =
+            OrderEventsCleanerConfig::new(Duration::from_millis(50), Duration::from_millis(200));
+        let cleaner = OrderEventsCleaner::new(config, db.clone());
+
+        tokio::task::spawn(cleaner.run_forever());
+
+        // delete `order_a` after the initialization
+        time::sleep(Duration::from_millis(20)).await;
```
Comment (fleupold) [inline crates/autopilot/src/periodic_db_cleanup.rs:126]: Can you comment in the code on why we are not using `tokio::time::advance` here (which is recommended for having more deterministic tests)? It may require using https://docs.rs/tokio/latest/tokio/time/struct.Instant.html instead of `Utc::now()` in the object under test, but otherwise I feel it should work?

**Review by fleupold — APPROVED** [review]
Looking good!

File: crates/autopilot/src/periodic_db_cleanup.rs (hunk 10 above)
Comment (squadgazzz) [reply crates/autopilot/src/periodic_db_cleanup.rs:133]: Added a comment
